File: lib/services/folder_service.dart
import 'package:flutter/foundation.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/folder_model.dart';

class FolderService {
  final SupabaseClient _supabase = Supabase.instance.client;

  Future<List<FolderModel>> getFolders({String? parentId}) async {
    try {
      final response = await _supabase
          .from('folders')
          .select()
          .eq('user_id', _supabase.auth.currentUser!.id)
          .order('created_at', ascending: false);
      
      if (response == null || response.isEmpty) {
        return [];
      }
      
      final allFolders = (response as List)
          .map((json) => FolderModel.fromJson(json as Map<String, dynamic>))
          .toList();
      
      if (parentId != null) {
        return allFolders.where((folder) => folder.parentId == parentId).toList();
      } else {
        return allFolders.where((folder) => folder.parentId == null).toList();
      }
    } catch (e) {
      debugPrint('Error getting folders: $e');
      return [];
    }
  }

  Future<FolderModel> createFolder(String name, {String? parentId}) async {
    try {
      final userId = _supabase.auth.currentUser!.id;
      
      final response = await _supabase.from('folders').insert({
        'user_id': userId,
        'name': name,
        'parent_id': parentId,
      }).select().single();

      return FolderModel.fromJson(response as Map<String, dynamic>);
    } catch (e) {
      debugPrint('Error creating folder: $e');
      rethrow;
    }
  }

  Future<void> deleteFolder(String folderId) async {
    await _supabase.from('folders').delete().eq('id', folderId);
  }

  Future<void> moveFileToFolder(String fileId, String? folderId) async {
    await _supabase
        .from('files')
        .update({'folder_id': folderId})
        .eq('id', fileId);
  }
}

File: services/folderService.ts
import { supabase } from '../config/supabase';
import { Folder } from '../types';

export const getFolders = async (parentId?: string | null): Promise<Folder[]> => {
  const { data: { user } } = await supabase.auth.getUser();
  if (!user) return []; // Return empty array instead of throwing

  let query = supabase
    .from('folders')
    .select('*')
    .eq('user_id', user.id)
    .order('created_at', { ascending: false });

  if (parentId) {
    query = query.eq('parent_id', parentId);
  } else {
    query = query.is('parent_id', null);
  }

  const { data, error } = await query;

  if (error) throw error;
  return (data || []) as Folder[];
};

export const createFolder = async (
  name: string,
  parentId?: string | null
): Promise<Folder> => {
  const { data: { user } } = await supabase.auth.getUser();
  if (!user) throw new Error('Not authenticated');

  const { data, error } = await supabase
    .from('folders')
    .insert({
      user_id: user.id,
      name,
      parent_id: parentId || null,
    })
    .select()
    .single();

  if (error) throw error;
  return data as Folder;
};

export const deleteFolder = async (folderId: string): Promise<void> => {
  const { error } = await supabase.from('folders').delete().eq('id', folderId);
  if (error) throw error;
};

export const moveFileToFolder = async (
  fileId: string,
  folderId: string | null
): Promise<void> => {
  const { error } = await supabase
    .from('files')
    .update({ folder_id: folderId })
    .eq('id', fileId);

  if (error) throw error;
};

